### DISCUSSION TITLE
pyinstaller 5.5 + tensorflow-macos-2.10 on M1 Pro

### DISCUSSION THREAD

**Discussion by odieXin** [pyinstaller/pyinstaller#7169] [Help] [answered]
Even if I have added paths, pyinstaller cannot find tensorflow and thus when I run the generated binary, it generated the below error:

```
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "tensorflow/__init__.py", line 29, in <module>
    from __future__ import absolute_import as _absolute_import
  File "_distutils_hack/__init__.py", line 88, in create_module
  File "importlib/__init__.py", line 127, in import_module
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "setuptools/__init__.py", line 10, in <module>
  File "_distutils_hack/__init__.py", line 88, in create_module
  File "importlib/__init__.py", line 127, in import_module
ModuleNotFoundError: No module named 'setuptools._distutils'
```

To reproduce it in a simpler environment, I created a test.py as below:
```
import tensorflow as tf
print (tf.__version__)
```
It runs well with "python3 test.py" 
But not, with "python3 -m PyInstaller test.py"

Pyinstaller 5.5
M1 Pro
Tensorflow-macos 2.10
Python 3.9

I have tried to add --paths, --hidden-hooks, --additional-hook-dir. None of these worked. 

Any help or suggested is highly appreciated

**Comment by rokm** [comment, marked as answer]
What version of `setuptools` do you have installed?

If you run the following snippets in your python environment, what is the output?
```python
from PyInstaller.utils.hooks import is_module_satisfies
print(is_module_satisfies("setuptools >= 60.0"))
```
```python
import setuptools
print(setuptools.__version__)
```

Does it help if you add `--hiddenimport setuptools._distutils` (and maybe also `--collect-submodules setuptools._distutils`) to your PyInstaller command?

**Reply by odieXin** [reply]
My current version is 58.0.4 
Do I need to upgrade to 60 and above to make it work?

**Reply by odieXin** [reply]
Adding both works!! Thanks a lot!! with the setuptools of 58.0.4

```
--hiddenimport setuptools._distutils 
--collect-submodules setuptools._distutils
```

**Reply by rokm** [reply]
Hmm, I'll take a look at 58.0.4 - I was under impression that `setuptools` started overriding `distutils` at 60.0, but it might have happened at an earlier version, in that case, the version checks in our hooks will need to be adjusted.

**Reply by rokm** [reply]
Okay, I've figured it out. `setuptools` indeed started overriding `distutils` from version 60.0 on, but the `_distutils_hack` has been around from 50.0. Before 50.0, the `distutils_precedence.pth` did not activate it by default (the default value for `SETUPTOOLS_USE_DISTUTILS` is "stdlib"), while from 60.0 on, it does (they changed the default value to "local").

Our runtime hook, which is used in lieu of `distutils_precedence.pth`, however, always enabled `_distutils_hack` - which means that if `setuptools` is collected (as is the case with `tensorflow`), we will trigger an error with `setuptools` >= 50.0 and < 60.0, because [the standard hook](https://github.com/pyinstaller/pyinstaller/blob/576bcf2d9fd75b7bc90237aedb6b69d6a3c04a5f/PyInstaller/hooks/hook-setuptools.py#L39-L41) forces collection of `setuptools._distutils` only for version >= 60.0.
 
We'll fix that in #7172.

**Reply by odieXin** [reply]
Thanks! @rokm
